### PR TITLE
nixos/nginx-sso: allow using file-based secrets

### PR DESCRIPTION
**Pull request by ambroisie**
## Description of changes

> This was living for a *long* time in my config, I finally decided to upstream it.

Also some very simple cleanups to the module, as it is somewhat showing its age.

I'd love to know if there is a better way to do this nowadays, I literally wrote this when I'd just started using NixOS three years ago.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ambroisie** [comment]
@ofborg test nginx-sso

File: nixos/modules/services/security/nginx-sso.nix [hunk 1]
```diff
@@ -47,14 +51,34 @@ in {
...
+        ExecStartPre = ''+${pkgs.writeShellScript "merge-nginx-sso-config" ''
+          rm -f '${configPath}'
+          # Relies on YAML being a superset of JSON
+          ${utils.genJqSecretsReplacementSnippet cfg.configuration configPath}
+
+          # Fix permissions
+          chown nginx-sso:nginx-sso ${configPath}
+          chmod 0600 ${configPath}
+        ''
+        }'';
```
Comment (SuperSandro2000) [inline nixos/modules/services/security/nginx-sso.nix:65]:
```suggestion
        ExecStartPre = pkgs.writeShellScript "merge-nginx-sso-config" ''
          rm -f '${configPath}'
          # Relies on YAML being a superset of JSON
          ${utils.genJqSecretsReplacementSnippet cfg.configuration configPath}
          # Fix permissions
          chown nginx-sso:nginx-sso ${configPath}
          chmod 0600 ${configPath}
        '';
```
Fixing permissions afterwards leaves us with a race condition where the permissions are not correct.

Comment (ambroisie) [reply nixos/modules/services/security/nginx-sso.nix:65]: The `+` in front of the script is load-bearing, it means it is executing as `root`.

Comment (ambroisie) [reply nixos/modules/services/security/nginx-sso.nix]: Added a note about the files needing to be readable by `nginx-sso`. We can remove the `+`.

File: nixos/modules/services/security/nginx-sso.nix [hunk 2]
```diff
@@ -37,6 +37,11 @@ in {
         nginx-sso configuration
         ([documentation](https://github.com/Luzifer/nginx-sso/wiki/Main-Configuration))
         as a Nix attribute set.
+
+        Options containing secret data should be set to an attribute set
+        containing the attribute `_secret` - a string pointing to a file
+        containing the value the option should be set to. This file should be
+        readable by `nginx-sso`.
```
Comment (SuperSandro2000) [inline nixos/modules/services/security/nginx-sso.nix:44]: Please add an example, alone from the description I don't understand it.
```suggestion
        Options containing secret data should be set to an attribute set
        with the attribute `_secret` which has a string assigned which points to a file
        containing the value that should be used in the end.
        This file must be readable by `nginx-sso`.
```

Comment (ambroisie) [reply nixos/modules/services/security/nginx-sso.nix:44]: I find that somewhat surprising, this is a know pattern, IIRC I had taken this description verbatim from GitLab's module.

I'm not a _big_ fan of your rephrasing, how about:
```suggestion
        Options containing secret data should be set to an attribute set
        with the singleton attribute `_secret` - a string value set to the path
        to the file containing the secret value which should be used in the
        configuration. This file should be readable by `nginx-sso`.
```

Comment (SuperSandro2000) [reply nixos/modules/services/security/nginx-sso.nix:44]: It should not be readable but nginx-sso, it must be, otherwise it won't work.

Comment (ambroisie) [reply nixos/modules/services/security/nginx-sso.nix:44]: Yeah I was also going to add an example, don't worry :-)

**Comment by ambroisie** [comment]
@SuperSandro2000 friendly ping :-)

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**